pkg/controller/control.go

```diff
@@ -1249,23 +1249,42 @@ func (c *Controller) handleDiskNoSpaceErrorForReplicas(replicaNoSpaceErrMap map[
 		return nil
 	}
 
-	replicaWritableCount := 0
+	replicaRWCount := 0
+	setNoSpaceRWReplicaErrList := make([]string, 0, len(c.replicas))
+	setNoSpaceWOReplicaErrList := make([]string, 0, len(c.replicas))
 	for _, r := range c.replicas {
-		if r.Mode == types.ERR {
+		if r.Mode == types.RW {
+			replicaRWCount++
+			if _, exist := replicaNoSpaceErrMap[r.Address]; exist {
+				setNoSpaceRWReplicaErrList = append(setNoSpaceRWReplicaErrList, r.Address)
+			}
+			continue
+		}
+		if r.Mode == types.WO {
+			if _, exist := replicaNoSpaceErrMap[r.Address]; exist {
+				setNoSpaceWOReplicaErrList = append(setNoSpaceWOReplicaErrList, r.Address)
+			}
 			continue
 		}
-		replicaWritableCount++
+	}
+
+	for _, address := range setNoSpaceWOReplicaErrList {
+		c.setReplicaModeNoLock(address, types.ERR)
 	}
 
 	// https://github.com/longhorn/longhorn/issues/10718#issuecomment-3190666485
 	// If a replica is on a disk that is out of space, and others are not, this replica will be marked as ERR to avoid the inconsistent data.
-	// The volume will be readable when all disks of the replicas are out of space so last one replica will not be marked as ERR because of ENOSPC.
-	if len(replicaNoSpaceErrMap) == replicaWritableCount {
+	// If all replicas are on disks that are out of space at the same time, the volume will be directly return an ENOSPC error and no replica will be set to ERR.
+	// Last one replica will not be marked as ERR for an ENOSPC error so the volume will still be readable when all disks of the replicas are out of space.
+	if len(setNoSpaceRWReplicaErrList) == replicaRWCount {
+		// the ErrNoSpaceLeftOnDevice error will be translated to TypeENOSPC/ENOSPC and handled by `liblonghorn`
 		return types.ErrNoSpaceLeftOnDevice
 	}
-	for address := range replicaNoSpaceErrMap {
+
+	for _, address := range setNoSpaceRWReplicaErrList {
```

Comment:
Why don't we set WO replicas to ERR?

Reply:
Modified. Directly set WO replicas to `ERR` when it is in `replicaNoSpaceErrMap`.